sound/soc/sof/intel/hda.c

```diff
@@ -472,10 +476,28 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	else
 		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
 
-	/* used by hda machine driver to create dai links */
-	mach_params = (struct snd_soc_acpi_mach_params *)
-		&sdev->pdata->machine->mach_params;
-	mach_params->codec_mask = bus->codec_mask;
+	if (bus->codec_mask) {
+		for (i = 0; i < HDA_MAX_CODECS; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/* if there are HDMI codec and HDA codecs, hda machine
+		 * is applied. For I2S codec + HDMI codec, there is only one
+		 * hda codec. For other cases, the value should be zero.
+		 */
+		if (codec_num > 1) {
+			/* used by hda machine driver to create dai links */
+			mach = snd_soc_acpi_intel_hda_machines;
+			pdata = sdev->pdata;
+			mach->pdata = pdata->machine->pdata;
+			mach->sof_fw_filename =
+					pdata->desc->nocodec_fw_filename;
+			mach_params = &mach->mach_params;
+			mach_params->codec_mask = bus->codec_mask;
+			pdata->machine = mach;
+		}
```

Comment:
do nocodec setup here:

else
        sof_nocodec_setup()

Reply:
please refer to sof_xxx_probe. machine driver carries some information for sof_probe.
And if no codec is found at probe function, a default codec should be used at there, no here

Reply:
@RanderWang as I see it the logic is:
if force-no-codec-mode
      use no-codec machine drive
else 
       If acpi match found 
             use the appropriate machine driver
      else
            use no-codec machine driver

if num-hda-codecs > 1
     use hda machine driver

Is this correct? then does this mean we cannot force no-code is hda is enabled?

Reply:
@ranj063  please try CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE.
if hda is enable and no hda codec, it would be nocodec